### ISSUE TITLE
Rung 7 round: deferred triage shapes on def-use (probe-first)

### ISSUE THREAD

**Issue by richlander** [richlander/dotnet-inspect#1941]
# Rung 7 round: deferred triage shapes on the def-use substrate (probe-first)

Rung 7 of the [Analysis/Facts bringup ladder (#1891)](https://github.com/richlander/dotnet-inspect/issues/1891). Re-attempt the triage shapes that were **parked as too noisy before def-use existed**, now that the substrate (reaching-defs, escape, EH-edges) is built. This is the "improve Performance Triage with more data" work, and it directly targets the Aspire dashboard's allocation profile (LINQ + async heavy). **0.16.0 ships after this round.**

## Philosophy (hard rules)

- **Probe before build.** Each candidate shape starts with a corpus measurement: how many candidate sites exist, and how cleanly def-use / R2 facts classify the actionable subset. **A documented decline is a valid, expected outcome** — these shapes were declined once for noise; we only ship what clears the bar.
- **Def-use gated.** Every shape gates on `ReachingDefinitions.IsComplete` (conservative when incomplete); the broad scan stays the recall net.
- **Strict-improvement corpus gate.** Cut over only on measured strict-improvement vs the refreshed baseline (precision gained and/or false positives dropped); no flat-corpus ships.
- **Type from metadata where possible** (per #1939): resolve allocation-site element/boxed types and sink types from operand tokens/signatures + reaching-defs — no typed-stack layer. If a shape genuinely can't get its type that way, that's a measured #1908 typed-stack trigger to report, not force.
- Two-family adversarial review (G5).

## Candidate shapes

- **`materialize-in-loop` hoistability** (incl. LINQ `ToList`/`ToArray`/`ToDictionary` and other copies). The original decline reason was "naive = noise." The def-use question: is the materialized value **loop-invariant** (hoistable out of the loop) vs genuinely recomputed per iteration? Reaching-defs answers it. Aspire-relevant (dashboard LINQ).
- **State-machine allocation** (async). The enumerator half landed (#1814); async state-machine `newobj` of compiler-generated `<...>d__N` async types is the deferred half. Recognize structurally and/or cross-fed from R2 iterator/async facts. Aspire-relevant (async-heavy dashboard).
- **Type-aware precision for existing shapes** (#1939 Layer-0 consumer): sharpen `small-array`/`stackalloc-candidate`/`box-value-type`/escape with operand-token element/boxed types + reaching-defs sink types. Lower-risk continuation of the escape-gate work.

### Track A — Probe (read-only, gates the build)

- **Goal:** quantify paydirt per candidate shape, so the build only targets shapes that clear the noise bar.
- **Owns:** AnalysisHarness / measurement scripts; **read-only on `LibraryBodyIndex`** (no edits) — parallel-safe with nothing.
- **Deliverable:** per shape — count of candidate sites on the corpus, the fraction cleanly classifiable via def-use / R2 facts (hoistable vs per-iteration; recognized async SM vs ambiguous), and a sampled precision read. A go/no-go recommendation per shape. Report on this issue.

### Track B — Build (owns `LibraryBodyIndex`)

- **Goal:** build the shapes Track A greenlights.
- **Owns (exclusive):** `LibraryBodyIndex` triage/opportunity paths. **One build-agent only** (guard G7 — no fan-out on the central file). Shapes land sequentially.
- **Per shape:** gate on `IsComplete`, two-tier (scan = recall net), resolve types from operand tokens/signatures where needed, validate against the refreshed corpus baseline, two-family review. Skip/decline any shape the probe shows is still noise.

## Guards

- One build-agent on `LibraryBodyIndex` (G7). Track A is read-only; Track B owns the edits.
- Refresh the corpus baseline if needed before Track B's strict-improvement diffs.
- After this round: **cut 0.16.0** with the new shapes + the precision wins already merged (cost/semantics overlays, escape gates, convergence).

Relates to #1891 (ladder, Rung 7), #1939 (type-aware triage as Layer-0 consumer), #1714 (static+dynamic pairing — later), #1805 (Aspire triage eval origin).

**Comment by richlander** [comment]
## Agent brief — Track A: Rung 7 probe (read-only, gates the build)

Quantify paydirt for each candidate shape **before** any `LibraryBodyIndex` edits, so Track B only builds shapes that clear the noise bar. These shapes were declined once for noise; the probe is the evidence gate. **Read-only — do not edit `LibraryBodyIndex`.** Work in AnalysisHarness / throwaway scripts; report numbers on #1941.

### Corpus
Use the standard analysis corpus (`eng/prepare-decompiler-corpus.sh` / the AnalysisHarness corpus list), and include **Aspire.Dashboard.dll** explicitly — it's the motivating target (LINQ + async heavy).

### Per shape, measure and report

**1. `materialize-in-loop` hoistability (incl. LINQ `ToList`/`ToArray`/`ToDictionary`)**
- Count call sites where a materialize/copy call occurs inside a loop region (back-edge heuristic).
- Of those, use **reaching-defs** to classify: is the materialized source **loop-invariant** (the collection/query def reaches the loop from outside, unchanged in-loop → hoistable, the actionable win) vs **recomputed per iteration** (depends on the loop variable → not hoistable, often correct as-is)?
- Report: total sites, hoistable count, per-iteration count, and `IsComplete`-incomplete count (skipped). Sample ~15 hoistable hits and label true/false positive by hand.
- **Go/no-go:** is the hoistable subset a meaningful, high-precision population, or noise?

**2. State-machine allocation (async)**
- Count `newobj` of compiler-generated async state-machine types (`<...>d__N` implementing `IAsyncStateMachine`, or the async builder allocation). Distinguish from the already-shipped enumerator shape (#1814).
- Classify: which are genuinely heap-allocated per call (class state machines / boxed structs on the async path) vs struct state machines the JIT keeps on the stack?
- Cross-check against R2 async/iterator recovery facts where available (don't re-derive structurally if the decompiler already recognizes it).
- Report: total async-SM allocation sites, the heap-allocating subset, and how cleanly they're recognizable on R1 vs needing R2 facts. **If recognition needs type/stack info R1 can't supply, that's a #1908 typed-stack-trigger candidate — flag it, don't force.**

**3. Type-aware precision for existing shapes (#1939 Layer-0)**
- For `box-value-type` / `small-array` / `stackalloc-candidate`: how often does resolving the **operand-token type** (boxed type, array element type) change the classification vs today's logic? Quantify the precision delta (false positives that a token-type check would drop, real positives currently missed).
- Report: estimated precision/recall change if the existing shapes used operand-token types + reaching-defs sink types.

### Deliverable
A go/no-go per shape with the numbers, plus a recommended build order for Track B (highest-paydirt first). **A documented decline for any shape that's still noise is a valid, expected outcome** — say so explicitly with the evidence. No code changes to ship from Track A.

**Comment by richlander** [comment]
## Track A amendment — use the #1939 (a)/(b)/(c) failure taxonomy for shape #3

For the **type-aware precision** measurement (shape #3), classify every type-resolution need using the [#1939 taxonomy](https://github.com/richlander/dotnet-inspect/issues/1939#issuecomment-4840146437), because only one cause is a real typed-stack trigger:

- **(a) on a token / signature / slot-def** — `newarr`/`box`/`newobj` operand tokens, `IsByRefLike` on the def, sink types on field/return tokens, alloc→sink via reaching-defs. **Build now** (token resolution). Count these.
- **(b)-reachability** — "is this alloc consumed by that call" with no spill to a local. A **stack-height** question (height tier / `ComputeMaxStack`-shaped), **no types**. Count these (they're build-now via heights, not typed stack).
- **(b)-residue** — type of an **untyped intermediate**, not on a token/def, surviving `dup`/multi-consumer juggling. **This is the only #1908 typed-stack trigger.** Count these precisely — this number is the whole point.
- **(c) cross-assembly `TypeRef` lost under no-load** — **NOT** a typed-stack trigger (the lattice coarsens `TypeRef`->`ObjRef` too). Fixable with a bounded **well-known-type table** (Span/ReadOnlySpan/ref-struct BCL types, unmanaged primitives, common boxed primitives) or accepted incompleteness. Count these and note which well-known types would resolve them.

**Deliverable addition:** a histogram of type-resolution needs across (a)/(b)-reachability/(b)-residue/(c) for the allocation/escape shapes. If **(b)-residue ≈ 0** (the prediction), the typed stack stays gated and the round proceeds entirely on token resolution + reaching-defs + a small well-known-type table — all build-now. If (b)-residue is non-trivial, that's the first *measured* typed-stack trigger and feeds the #1908 decision.

**Comment by richlander** [comment]
## Track A probe report — Rung 7 deferred triage shapes

Read-only probe over the standard analysis/decompiler corpus plus Aspire Dashboard:

- corpus list: `eng/prepare-decompiler-corpus.sh /tmp/issue1941-corpus.txt`
- explicit Aspire add: `/home/rich/git/aspire-perf-triage/artifacts/bin/Aspire.Dashboard/Release/net8.0/Aspire.Dashboard.dll`
- probe script: file-based app in `/tmp/rung7probe.cs` using `ILInspector.Analysis` public APIs plus raw metadata/IL reads
- build-event validation: `build-event-agent.sh validate tools/AnalysisHarness -- -c Release --nologo --verbosity quiet`
  - Summary: 30 projects, 0 failed, 0 errors, 0 warnings
  - Event log: `20260630T052623.4997886Z-2241097-build-1195d597`

## 1. Materialize-in-loop hoistability

Measured call/copy sites in loop regions for `ToArray`, `ToList`, `ToDictionary`, `ToHashSet`, plus string/copy-like `Substring` and `ToString`.

```text
total=484
hoistable=130
per-iteration=53
unknown=301
```

Breakdown by callee:

```text
Substring   hoistable=17  per-iteration=21  unknown=61
ToArray     hoistable=38  per-iteration=4   unknown=24
ToDictionary hoistable=0  per-iteration=0   unknown=3
ToHashSet   hoistable=0   per-iteration=0   unknown=3
ToList      hoistable=12  per-iteration=11  unknown=68
ToString    hoistable=63  per-iteration=17  unknown=142
```

Aspire Dashboard:

```text
Aspire.Dashboard.dll total=44 hoistable=2 perIteration=4 incomplete=0 unknown=38
```

Sample hoistable LINQ/materializer rows:

```text
Newtonsoft.Json.dll    BinaryConverter::ReadByteArray                    IL_0037  ToArray  source local defined before loop
Microsoft.CodeAnalysis.dll SwitchStringJumpTableEmitter::EmitHashTableSwitch IL_004F ToArray source argument reaches loop
DotnetInspector.Services.dll GitHubUrlResolver::ResolveSampleUrl         IL_00A4  ToList   source local defined before loop
dotnet-inspect.dll     MemberCodeViewMarkoutTypeInfo::Serialize          IL_0426  ToArray  source argument reaches loop
```

Read: the def-use discriminator is useful, but the broad candidate family is still noisy. `ToString` / `Substring` rows dominate and need a separate taste/gate discussion; many are not actionable "materialize" wins. The LINQ/materializer subset (`ToArray` / `ToList`) has a real hoistable population, but unknown remains high.

Recommendation:

- **Go, narrowly**, for Track B only on `ToArray` / `ToList` materializers where the source is cleanly classified by reaching-defs as loop-invariant.
- **Decline for now** on broad `ToString` / `Substring` copy rows; they are too noisy for this round.
- Keep unknown rows out of the shipped shape; `IsComplete`/simple-source classification should gate the build.

## 2. Async state-machine allocation

Current `AllocationOccurrence` does not surface async state-machine allocations (`AllocationKind.StateMachine` async subset was 0), but raw metadata found async state-machine types and direct class-state-machine allocations.

Corpus raw metadata summary:

```text
Newtonsoft.Json.dll                 asyncTypes=89  structTypes=89  classTypes=0   newobj=0
Microsoft.CodeAnalysis.dll          asyncTypes=78  structTypes=78  classTypes=0   newobj=0
Microsoft.CodeAnalysis.CSharp.dll   asyncTypes=1   structTypes=1   classTypes=0   newobj=0
System.CommandLine.dll              asyncTypes=4   structTypes=4   classTypes=0   newobj=0
Microsoft.ApplicationInsights.dll   asyncTypes=14  structTypes=14  classTypes=0   newobj=0
DotnetInspector.Core.dll            asyncTypes=2   structTypes=1   classTypes=1   newobj=0
DotnetInspector.Packages.dll        asyncTypes=30  structTypes=30  classTypes=0   newobj=0
DotnetInspector.Services.dll        asyncTypes=13  structTypes=13  classTypes=0   newobj=0
dotnet-inspect.dll                  asyncTypes=105 structTypes=105 classTypes=0   newobj=0
Aspire.Dashboard.dll                asyncTypes=458 structTypes=448 classTypes=10  newobj=16 classNewobj=16
```

Aspire class-state-machine allocation sites:

```text
Aspire.Dashboard.ServiceClient.DashboardClient::SubscribeConsoleLogs       IL_0002 <SubscribeConsoleLogs>d__71 class
Aspire.Dashboard.ServiceClient.DashboardClient::GetConsoleLogs             IL_0002 <GetConsoleLogs>d__72 class
Aspire.Dashboard.Otlp.Storage.TelemetryRepository::WatchSpansAsync         IL_0002 <WatchSpansAsync>d__124 class
Aspire.Dashboard.Otlp.Storage.TelemetryRepository::WatchLogsAsync          IL_0002 <WatchLogsAsync>d__125 class
Aspire.Dashboard.Api.TelemetryApiService::FollowSpansAsync                 IL_0002 <FollowSpansAsync>d__9 class
Aspire.Dashboard.Api.TelemetryApiService::FollowLogsAsync                  IL_0002 <FollowLogsAsync>d__10 class
... plus GetAsyncEnumerator allocations for those async enumerable classes
```

Read: this is a clean Aspire-motivated population. R1 can identify the type by metadata (`IAsyncStateMachine` plus class vs value type) and the allocation site by `newobj`. No typed stack needed for the class-state-machine subset.

Recommendation:

- **Go** for Track B on class async state-machine `newobj` sites.
- **Do not report struct async state machines** unless a concrete boxing/allocation site is observed; the probe saw many struct async state-machine types but no struct `newobj` allocation sites.
- This should be the first build target: it is precise, Aspire-relevant, and does not require typed stack.

## 3. Type-aware precision taxonomy (#1939 buckets)

Rubric used:

- `(a)` token/signature/slot-def
- `(b)-reachability` stack-height/reachability, no type
- `(b)-residue` untyped intermediate; only this is a typed-stack trigger
- `(c)` cross-assembly `TypeRef` / well-known table gap

Histogram across current allocation/escape shapes:

```text
total=16724
(a)=16724
(b)-reachability=0
(b)-residue=0
(c)=0
```

By shape:

```text
array           (a)=5260
box             (a)=5894
box-value-type  (a)=1283
small-array     (a)=4287
```

Aspire Dashboard:

```text
Aspire.Dashboard.dll total=2292 (a)=2292 reachability=0 residue=0 c=0
```

Read: the measured set is entirely build-now `(a)` token/signature work. There were **zero** `(b)-residue` rows, so this probe found no typed-stack trigger for the shipping allocation/escape triage shapes. There were also zero `(c)` rows in the measured set; no well-known-type table need surfaced in this corpus slice.

Recommendation:

- **Go** for Track B type-aware precision work using token/signature/slot-def evidence.
- Keep typed stack gated. This probe found no measured `(b)-residue` trigger.
- If future work adds shapes that rely on untyped stack intermediates, rerun the taxonomy; only nonzero `(b)-residue` should reopen #1908 typed-stack discussion.

## Recommended Track B order

1. **Async class state-machine allocation** — highest Aspire paydirt, clean metadata discriminator, no typed-stack dependency.
2. **Type-aware precision for existing shapes** — all measured needs are `(a)`, so this is build-now and low risk.
3. **Narrow LINQ materialize-in-loop hoistability** — build only the def-use-clean `ToArray` / `ToList` subset; leave string/copy rows and unknown rows out of this round.

## Build-event workflow note

Used validation-only mode because Track A's primary proof is probe evidence, not build diagnostics. The build-event views were green and supplemental:

```text
summary: projects=30 failed=0 errors=0 warnings=0
scorecard: green, DiagnosticTypes=0
```

**Comment by richlander** [comment]
## Probe review — greenlit, with build order and one caveat on the async shape

Strong, honest measurement. The go/no-go calls are well-gated and the decline-where-noisy discipline held. Greenlighting Track B in the recommended order, with one refinement on shape #1.

### The headline result: typed stack stays gated, now on *evidence*

`(a)=16724, (b)-reachability=0, (b)-residue=0, (c)=0` across the shipping allocation/escape shapes is the measured answer we've been after: **no typed-stack trigger exists for any current shape.** This closes the #1908 question with data, not prediction — exactly matching the checkpoint and the #1939 reviewers. Type-aware precision is 100% build-now `(a)` token/signature work. (Note `(c)=0` is corpus-slice-specific — the allocation shapes resolve types from the alloc-site operand token, which doesn't hit the cross-assembly `TypeRef` gap; that gap lives in decompiler rendering, not these shapes. Re-measure if a shape ever reaches for a referenced-assembly type.)

### Build order — agreed

1. **Async class state-machine allocation** — first. Clean metadata discriminator (`IAsyncStateMachine` + class vs struct), R1-only, and genuinely Aspire-motivated (the 10 class SMs are the dashboard's streaming/subscription methods — `WatchSpansAsync`, `FollowLogsAsync`, `SubscribeConsoleLogs`...). Struct-SM correctly excluded (no heap `newobj` observed).
2. **Type-aware precision** — build-now, all `(a)`, low risk.
3. **Narrow LINQ materialize** — only the def-use-clean `ToArray`/`ToList` loop-invariant subset; decline the noisy `ToString`/`Substring` family and the `unknown` rows. Accept low recall for high precision (2/44 clean on Aspire is fine — precision over recall).

### Caveat on shape #1 (actionability/frequency — the eliminated-vs-moved point)

The class async state machines you found are almost all **async iterators** (`IAsyncEnumerable` — the streaming methods), which allocate the class **once per enumeration**, not per item. For a long-lived dashboard subscription that's **amortized**, not a hot steady-state allocation — the same category as the constructor allocations we already demote. So the shape must carry **frequency/actionability context**, or it will flag Aspire's streaming methods as if they were hot per-call allocations:

- Mark once-per-enumeration / once-per-subscription async-SM allocations as **amortized** (use the loop/caller-graph context the tool already has — per-call-in-a-loop is the hot case; once-per-stream is amortized).
- This is exactly #1714's **"eliminated vs moved"** discipline: an async SM is the canonical *moved-not-eliminated* allocation. The `SafeFixDirection`/caveat should say "intrinsic to async streaming; optimize only if this is allocated in a hot loop, not once per stream" — not imply it's a free fix.

Without that, shape #1 over-flags the very Aspire methods that motivated it. With it, it's a precise, honest, Aspire-relevant win. That nuance is the should-fix for Track B's first build.

### Net
Go on all three in the stated order; ship the async-SM shape **with amortization/frequency context**; keep the materialize subset narrow; type-aware precision is pure `(a)` build-now. Typed stack stays gated on the measured `(b)-residue=0`. Refresh the corpus baseline before the strict-improvement diffs, two-family review per shape.
